### PR TITLE
Add email sign in - Part 1

### PR DESCRIPTION
**Pull request by rohitpaulk**
Part of https://github.com/gratipay/gratipay.com/issues/1052. https://github.com/gratipay/gratipay.com/issues/1052#issuecomment-309198597, specifically.

This adds a table to store email auth nonces + a few helper functions to create, verify and invalidate records in that table.

Controllers etc. in next PR (#4539). This can be merged and deployed independently of those changes.

**Non-critical items to be reticketed:** 

- [ ] Add cron to clean out old `email_auth_nonces`

### CODE REVIEW

**Comment by rohitpaulk** [comment]
Ready for review

File: gratipay/models/participant/email.py [hunk 1]
```diff
@@ -54,6 +54,20 @@ class Email(object):
 
     """
 
+    @classmethod
+    def from_email(cls, email_address):
```
Comment (whit537) [inline gratipay/models/participant/email.py]: I would expect this be in the __init__.py with the other constructors. No?

File: gratipay/security/authentication/website_helpers.py [hunk 2]
```diff
@@ -86,3 +80,8 @@ def add_auth_to_response(response, request=None, user=ANON):
     if SESSION in request.headers.cookie:
         if not user.ANON:
             user.keep_signed_in(response.headers.cookie)
+
+def start_user_as_anon():
+    """Make sure we always have a user object, regardless of exceptions during authentication.
+    """
+    return {'user': ANON}
```
Comment (whit537) [inline gratipay/security/authentication/website_helpers.py]: 👍

File: sql/branch.sql [hunk 3]
```diff
@@ -0,0 +1,16 @@
+BEGIN;
+    CREATE TYPE email_auth_intent AS ENUM ('sign-in');
```
Comment (whit537) [inline sql/branch.sql]: What are you thinking the other intents will be?

File: sql/branch.sql [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+    CREATE TABLE email_auth_nonces
+    ( id                    serial                      PRIMARY KEY
+    , email_address         text                        NOT NULL
```
Comment (whit537) [inline sql/branch.sql]: Doesn't this duplicate `emails.address`?

File: sql/branch.sql [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    , email_id              integer                     REFERENCES emails
+                                                          ON UPDATE CASCADE
+                                                          ON DELETE CASCADE
+    , intent                email_auth_intent           NOT NULL
+    , nonce                 text                        NOT NULL
+    , ctime                 timestamp with time zone    NOT NULL DEFAULT CURRENT_TIMESTAMP
+    , UNIQUE (email_address, nonce)
```
Comment (whit537) [inline sql/branch.sql]: Since we're using UUID for the nonce, `UNIQUE(nonce)` should be enough, no?

**Comment by whit537** [comment]
Made some drive-by comments on first skim. Will go deeper on next pass ...

File: gratipay/models/participant/email.py (hunk 1 above)
Comment (rohitpaulk) [inline gratipay/models/participant/email.py]: Hmm, yh probably better placed there.

File: sql/branch.sql (hunk 3 above)
Comment (rohitpaulk) [inline sql/branch.sql]: `sign-up`

Comment (rohitpaulk) [inline sql/branch.sql]: The purpose of storing this is that for `sign-in`, we have to make sure that an email reference (to the `emails` table) is maintained. When an email is deleted, this record should be deleted too. (Which reminds me, got to add a test for that) This prevents the case where a user mistakenly adds an email address, and an attacker verifies it and uses it to log in. (Not sure how likely it is, but plausible)

File: sql/branch.sql (hunk 4 above)
Comment (rohitpaulk) [inline sql/branch.sql]: We won't always have a reference to the `emails` table, this is only for `sign-in`. (check the `sign_in_has_valid_email` constraint below).

File: sql/branch.sql (hunk 5 above)
Comment (rohitpaulk) [inline sql/branch.sql]: Yes, should be. Will amend

Comment (rohitpaulk) [inline sql/branch.sql]: Fixed.

File: gratipay/models/participant/email.py (hunk 1 above)
Comment (rohitpaulk) [inline gratipay/models/participant/email.py]: Fixed.

File: sql/branch.sql (hunk 3 above)
Comment (rohitpaulk) [inline sql/branch.sql]: > Which reminds me, got to add a test for that

Added.

Comment (rohitpaulk) [inline sql/branch.sql]: > The purpose of storing this is that for sign-in, we have to make sure that an email reference (to the emails table) is maintained. When an email is deleted, this record should be deleted too. (Which reminds me, got to add a test for that) This prevents the case where a user mistakenly adds an email address, and an attacker verifies it and uses it to log in soon after the user has deleted it. (Not sure how likely it is, but plausible)

Just realized that we don't need this. When the email link is hit, we'll have to verify that the email (and linked participant) exists anyway. Removing.

Comment (rohitpaulk) [inline sql/branch.sql]: Removed.

File: sql/branch.sql (hunk 4 above)
Comment (rohitpaulk) [inline sql/branch.sql]: Note: Removed the reference to `emails`, not needed

**Comment by whit537** [comment]
@rohitpaulk What do you think of 1e5ada688a5c7a389a5d22fc9aeba322d2462034? The API seemed parallel enough to include there instead of under `gratipay/security`.

**Comment by whit537** [comment]
Honestly I'd be more comfortable integrating this and downstream work in a feature branch before merging to master. IMO that worked pretty well on #4488. It gives us a chance to iterate on the API more freely than if we are deploying to master at each step (it also avoids the feature-flag complication). We may decide to vary the API or even the database schema based on downstream considerations. You willing to give that a try, @rohitpaulk?

**Comment by whit537** [comment]
If so, we can simply leave this open as our integration branch/PR, and merge #4539 and additional work into here before finally landing.

**Comment by rohitpaulk** [comment]
> What do you think of 1e5ada6? The API seemed parallel enough to include there instead of under gratipay/security.

Hmm, that might be a problem when implementing sign up. The decoupling was intentional, we won't always have a `Participant` object available when dealing with nonces.

**Comment by rohitpaulk** [comment]
> Honestly I'd be more comfortable integrating this and downstream work in a feature branch before merging to master. IMO that worked pretty well on #4488. It gives us a chance to iterate on the API more freely than if we are deploying to master at each step (it also avoids the feature-flag complication). We may decide to vary the API or even the database schema based on downstream considerations. You willing to give that a try, @rohitpaulk?

#4488 was merged in a hurry before sustain, not the best example IMO. Making changes to the API was hard then too (even though we used layered PRs and didn't merge stuff early), and we decided to reticket [a major change](https://github.com/gratipay/gratipay.com/issues/4520). 

I understand how this can be difficult to review though. Half of the work that builds upon this is up in my head and not documented anywhere. I'll work on sign-up a bit before resurfacing this for review. This change is currently at ~500 lines modified, including both PRs. I'll aim for a cap of ~1000 lines modified. If it goes past that, let's find a way to merge a safe enough portion in early? It is difficult to maintain multiple PRs and >1000 lines outstanding.

Closing this PR, keeping the other open.

**Comment by whit537** [comment]
Okay, let's see how that works.

**Pull request closed without merging**